### PR TITLE
add: pytorch implementation

### PR DESCRIPTION
**Pull request by cibeah**
This is a PR to add a PyTorch implementation of AGAC. 
The source code and `README` to launch the scripts have been added separately in a new `agac_torch` folder.

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -0,0 +1,19 @@
+repos:
```
Comment (yfletberliac) [inline .pre-commit-config.yaml:1]: What is the purpose of using pre-commit here? (Naive question)

File: agac_torch/configs/point_maze.yaml [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    "adversary_learning_rate": 0.00009
+    "layers_dim": [64, 64]
+    "adv_layers_dim": [64, 64]
+    "discount": 0.99
+    "batch_size": 512
+    "lambda_gae": 0.95
+    "intrinsic_reward_coefficient": 0.0004
+    "entropy_coefficient": 0.0
+    "clipping_epsilon": 0.2
+    "value_loss_clip": 1000
```
Comment (yfletberliac) [inline agac_torch/configs/point_maze.yaml:29]: `value_loss_clip=1000` means no clip, right?

File: agac_torch/configs/point_maze.yaml [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    "value_loss_coeff": 0.25
```
Comment (yfletberliac) [inline agac_torch/configs/point_maze.yaml:30]: Usually `value_loss_coeff` is set to 0.5: have you observed some changes in the performance with `value_loss_coeff=0.25`? (I suspect it does not affect the perf because it is just a constant in the gradient update, but it's fine to keep it as is)

File: agac_torch/requirements.txt [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+gym~=0.18.0
+gym-minigrid==1.0.1
+joblib==0.16.0
+lxml==4.5.2
+mpi4py==3.0.3
+mujoco-py==2.0.2.11
+multidict==5.0.0
+neptune-client==0.9.8
+numpy~=1.18.5
+pre-commit==2.9.2
```
Comment (yfletberliac) [inline agac_torch/requirements.txt:11]: Same (naive) question: what is the use of pre-commit here?

**Review by yfletberliac — COMMENTED** [review]
Dear @cibeah,

This really is a fantastic PR, and it seems like a huge amount of work to produce such clean and complete code. The parallelization via MPI, logging and visualization features is a big addition. We are very grateful, thank you.

Overall, I do not see any critical change to be made to the code, and the inclusion in the original repo is fine as it is. For clarity and ease of navigation in the repo, I'll just add a link to your `agac_torch` version in the root `README` of the repo. Else than that, I have only listed some questions which you will find in the review.

Best,
Yannis

File: .pre-commit-config.yaml (hunk 1 above)
Comment (cibeah) [reply .pre-commit-config.yaml:1]: We use pre-commit to automatically scan (and fix when possible) some code lines contravening PEP-8's guidelines (like sorting imports, lines that are too long, etc.). I configured it so it would only apply to the agac_torch/ subfolder !

File: agac_torch/configs/point_maze.yaml (hunk 2 above)
Comment (cibeah) [reply agac_torch/configs/point_maze.yaml:29]: Yes you're right ! No the cleanest way to implement it

File: agac_torch/configs/point_maze.yaml (hunk 3 above)
Comment (cibeah) [reply agac_torch/configs/point_maze.yaml:30]: It's actually the same thing if I'm not mistaken. The original implem has 
`self.vf_loss = .5 * tf.reduce_mean(tf.maximum(vf_losses1, vf_losses2))` 
and then 
`loss = self.pg_loss  -  [...]  + self.vf_coef * self.vf_loss [...]`.
I didn't add the `0.5` in the torch version and thought I might as well collapse both coefs in the `value_loss_coeff`:
`value_loss = self._value_loss_coeff * torch.max(vloss1, vloss2).mean()`

File: agac_torch/requirements.txt (hunk 4 above)
Comment (cibeah) [reply agac_torch/requirements.txt:11]: Ditto !

File: .pre-commit-config.yaml (hunk 1 above)
Comment (yfletberliac) [reply .pre-commit-config.yaml:1]: Got it, thanks for the explanation!

File: agac_torch/configs/point_maze.yaml (hunk 3 above)
Comment (yfletberliac) [reply agac_torch/configs/point_maze.yaml:30]: Ok! Makes sense.
